### DISCUSSION TITLE
What is the best way to recolor PNG files with alpha? Examples in post.

### DISCUSSION THREAD

**Discussion by futuremotiondev** [ImageMagick/ImageMagick#5572] [Help] [answered]
I am trying to write a PowerShell script that will invoke IM to re-color single color png files. Here is an image that visually demonstrates what I am trying to do:

![IMExample](https://user-images.githubusercontent.com/33441569/190927331-7be52104-f7c1-438d-b8ab-74da8d6c9085.png)

The image above only shows the destination color as grey but I want to be able to control that in my script. 

I've looked at [this](https://imagemagick.org/discourse-server/viewtopic.php?f=1&t=34834) and [this](https://stackoverflow.com/questions/35554282/how-to-colorize-a-black-transparent-png-icon-with-imagemagick), but I'm unsure what route to take. 

Can someone suggest the best command to delegate to IM in order to achieve what I want?

I've attached a zip of the original PNGs that I want to color.

Thanks so much for any help.

[PNG Testing.zip](https://github.com/ImageMagick/ImageMagick/files/9595209/PNG.Testing.zip)

**Comment by fmw42** [comment]
What is your IM version?

**Comment by futuremotiondev** [comment]
ImageMagick 7.1.0-47 Q16-HDRI x64 15861e0:20220827

**Comment by futuremotiondev** [comment]
Also, I just tried a command I found online, but it doesn't work:

magick in.png -channel RGB -fill #FF0000 -opaque #000000 -alpha on out.png

It just gives me the same png without any modified colors.

**Comment by fmw42** [comment]
Try this:

![Actions_About32x32](https://user-images.githubusercontent.com/13458218/190928195-75435f3e-9a95-4588-9521-36ba528f1de1.png)

 `magick Actions_About32x32.png -alpha deactivate -colorspace gray -auto-level +level-colors red,black -alpha activate result.png`

![result](https://user-images.githubusercontent.com/13458218/190928204-c87ca45d-81c2-4743-969a-809bfb165617.png)

Change red to any color you want.

**Reply by futuremotiondev** [reply]
It seems to work in Explorer, but the PNG is all messed up when opened in Photoshop. See this:

![Photoshop_bHAzBwCNqI](https://user-images.githubusercontent.com/33441569/190928413-133d5035-1d6f-4682-be15-a60c78d49664.png)

![Photoshop_E0WJvJPhFl](https://user-images.githubusercontent.com/33441569/190928497-98452d49-c9f7-4a85-b782-805319137d47.png)


Any ideas?

**Comment by futuremotiondev** [comment]
Another update. If I convert the result PNG file to .TIF just doing `magick result.png result.tiff` it creates a file that can be opened properly in photoshop. But if I convert it back to png with `magick result.tiff result.png` it creates an invalid PNG file. See this screen recording:

https://user-images.githubusercontent.com/33441569/190928679-8bd730b6-2dd7-4359-9e4a-328a5b45501d.mp4

**Reply by fmw42** [reply]
Define invalid!  in what way is it wrong?  What is happening?

**Reply by futuremotiondev** [reply]
Look at the screen recording. This is what results:

![Photoshop_UbpXU8JNb4](https://user-images.githubusercontent.com/33441569/190928813-be42f991-04cb-4f24-895f-c3201dab5521.png)

Doesn't look valid to me.

**Reply by fmw42** [reply]
When I open the input and resulting png files in PS, they look fine. So again what is wrong

**Reply by futuremotiondev** [reply]
I think the problem has something to do with indexed colors. When I recolor one of the original PNGs to red and then open it in Photoshop, I get this:

![Photoshop_0eArp1q2qx](https://user-images.githubusercontent.com/33441569/190928911-30e49b5d-f9b8-4872-9a1a-b1035bc3d379.png)

Photoshop says the image is in indexed mode. It should be 8/16bit RGBA. Is there a way to force the conversion to not use indexed color mode?

**Reply by futuremotiondev** [reply]
> When I open the input and resulting png files in PS, they look fine. So again what is wrong

They don't look fine on my end. What version of Photoshop are you using?

**Reply by fmw42** [reply]
Here are my screen shot images. They look fine to me.  Mac OSX Monterey, PS 23.5.1 


![Screen Shot 2022-09-18 at 2 34 37 PM](https://user-images.githubusercontent.com/13458218/190928963-98017842-e096-4f4a-b805-9742fa93d8b0.png)




![Screen Shot 2022-09-18 at 2 34 47 PM](https://user-images.githubusercontent.com/13458218/190928966-2fa3e675-a085-4650-8168-f7666ad551eb.png)

**Comment by fmw42** [comment, marked as answer]
To force to 32-bits color add PNG32:result.png in place of result.png

**Reply by futuremotiondev** [reply]
That fixed it. Thank you. My TIF conversion script is still busted though. I'll have to open another thread for that.

**Reply by fmw42** [reply]
If you are saving a transparent tif, you may need to specify the -define tiff:alpha=associated|unassociated|unspecified. I do not recall which one, but try unspecified.
